[DOCS] OpenTelemetry - add OpenTelemetry protocol support content (#4964)
elastic/apm-server: docs/guide/opentelemetry-elastic.asciidoc

> // Make tab-widgets work

Reviewer: The `:ot-pipelines:` link on line 7 needs to be updated to `https://opentelemetry.io/docs/collector/configuration/#service`

Author: Updated.

OpenTelemetry project, see the {ot-spec}[spec].

Elastic OpenTelemetry integrations allow you to reuse your existing OpenTelemetry
instrumentation to quickly analyze distributed traces and metrics to help you monitor
business KPIs and technical components with the {stack}.

There are two Elastic OpenTelemetry integrations available:

* <<open-telemetry-elastic-exporter,Elastic exporter on the OpenTelemetry collector>> (recommmended)
* <<open-telemetry-elastic-protocol,APM Server native support of OpenTelemetry protocol>> (experimental)

[[open-telemetry-elastic-exporter]]
==== Elastic exporter on the OpenTelemetry collector

The is the recommended OpenTelemetry integration. We have extended the "contrib" OpenTelemetry collector by
adding an Elastic exporter so that you can drop this integration into your current OpenTelemetry setup.

The architecture consists of three main components.

image::images/open-telemetry-exporter-arch.png[OpenTelemetry Elastic exporter architecture diagram]

|=== 

| *Agents* | The OpenTelemetry agents instrument the applications and export the telemetry data to the OpenTelemetry collector.

> | *OpenTelemetry collector* | The https://opentelemetry.io/docs/collector/configuration/#a-namereceiversaimg-width35-srchttpsrawgithubcomopen-telemetryopentelemetryiomainiconography32x32receiverssvgimg-receivers[receiver]
> collects the telemetry data from the OpenTelemetry agent, and then the https://opentelemetry.io/docs/collector/configuration/#a-nameprocessorsaimg-width35-srchttpsrawgithubcomopen-telemetryopentelemetryiomainiconography32x32processorssvgimg-processors[processor]

Reviewer: Fixing anchors

```suggestion
| *OpenTelemetry collector* | The https://opentelemetry.io/docs/collector/configuration/receivers[receiver]
collects the telemetry data from the OpenTelemetry agent, and then the https://opentelemetry.io/docs/collector/configuration/#processors[processor]
```

Author: Great. Thanks!

defines optional transformations on the data before it's exported using the Elastic exporter.

| *Elastic exporter* | The exporter translates the OpenTelemetry data collected from your services, applications, and infrastructure to Elastic's protocol.
The data includes trace data and metrics data. By extending the OpenTelemetry collector, no changes are needed in your instrumented services to begin using the {stack}.

|===

[float]
[[open-telemetry-collector-config]]
===== Download and configure the collector

> OpenTelemetry Collectors can be run an agents or as standalone collectors.

Reviewer: ```suggestion
OpenTelemetry Collectors can be run as agents or as standalone collectors.
```

There are two Elastic OpenTelemetry integrations available:

* <<open-telemetry-elastic-exporter,Elastic exporter on the OpenTelemetry collector>> (recommmended)
* <<open-telemetry-elastic-protocol,APM Server native support of OpenTelemetry protocol>> (experimental)

[[open-telemetry-elastic-exporter]]
==== Elastic exporter on the OpenTelemetry collector

The recommended OpenTelemetry integration is using the Elastic exporter on the OpenTelemetry collector.
We have extended the "contrib" OpenTelemetry collector by adding an Elastic exporter so that you can drop this
integration into your current OpenTelemetry setup.

The architecture consists of three main components.

image::images/open-telemetry-exporter-arch.png[OpenTelemetry Elastic exporter architecture diagram]

|=== 

| *Agents* | The OpenTelemetry agents instrument the applications and exports the telemetry data to the OpenTelemetry collector.

| *OpenTelemetry collector* | Collects the telemetry data (https://opentelemetry.io/docs/collector/configuration/#a-namereceiversaimg-width35-srchttpsrawgithubcomopen-telemetryopentelemetryiomainiconography32x32receiverssvgimg-receivers[receivers])

> from the OpenTelemetry agent, defines optional transformations (https://opentelemetry.io/docs/collector/configuration/#a-nameprocessorsaimg-width35-srchttpsrawgithubcomopen-telemetryopentelemetryiomainiconography32x32processorssvgimg-processors[processors])

Author: Good point. It's a little confusing. I have rephrased the sentence. Hopefully it's a lot clearer.

TIP: We recommend using {metricbeat-ref}/metricbeat-overview.html[{metricbeat}] to get a mature collector with more integrations
and integrated visualizations to collect infrastructure metrics.

[float]
[[open-telemetry-elastic-config]]
===== Elastic exporter configuration options

|=== 

| `apm_server_url` | Elastic APM Server URL. (required).

| `api_key` | Credential for {apm-server-ref-v}/api-key.html[API key authorization]. Must also be enabled in Elastic APM Server. (optional)

| `secret_token` | Credential for {apm-server-ref-v}/secret-token.html[secret token authorization]. Must also be enabled in Elastic APM Server. (optional)

| `ca_file` | Root Certificate Authority (CA) certificate for verifying the server's identity if TLS is enabled. (optional)

| `cert_file` | Client TLS certificate. (optional)

| `key_file` | Client TLS key. (optional)

| `insecure` | Disable verification of the server's identity if TLS is enabled. (optional)

|===

[float]
[[instrument-apps-collector]]
===== Instrument applications

To export traces and metrics to the OpenTelemetry Collector, ensure that you have instrumented your services and applications

> with the OpenTelemetry API, SDK, or both. For example, if you are a Java developer, you need to execute your Java app using the

Reviewer: What does execute mean in this context?

```suggestion
with the OpenTelemetry API, SDK, or both. For example, if you are a Java developer, you need to instrument your Java app using the
```

https://github.com/open-telemetry/opentelemetry-java-instrumentation[OpenTelemetry agent for Java]. 

By defining the following environment variables, you can customize the OTLP endpoint the agent will use to communicate with
APM Server.

[source,bash]
----

> export OTEL_RESOURCE_ATTRIBUTES=service.name=frontend,service.version=1.0-SNAPSHOT,deployment.environment=staging

Reviewer 2: `service.version=1.0-SNAPSHOT` -> `service.version=1.1.0`

export OTEL_EXPORTER_OTLP_ENDPOINT=https://apm_server_url:8200
java -javaagent:/path/to/opentelemetry-javaagent-all.jar \
     -jar target/frontend-1.1.jar
----

|=== 

| `OTEL_RESOURCE_ATTRIBUTES` | The service name to identify your application.

| `OTEL_EXPORTER_OTLP_ENDPOINT` | APM Server URL. The host and port that APM Server listens for events on.

|===

> You are now ready to collect <<open-telemetry-elastic-traces-metrics,traces and metrics>>, <<open-telemetry-elastic-verify,verifying metrics>>,
> and <<open-telemetry-elastic-kibana,visualizing metrics>> in {kib}.

Reviewer: Maybe remove these gerunds.

```suggestion
You are now ready to collect <<open-telemetry-elastic-traces-metrics,traces and metrics>>, <<open-telemetry-elastic-verify,verify metrics>>,
and <<open-telemetry-elastic-kibana,visualize metrics>> in {kib}.
```

[[open-telemetry-elastic-protocol]]
==== APM Server native support of OpenTelemetry protocol

experimental::[This feature is experimental and may be changed in a future release. It is only available in a self-managed environment.]

> The APM Server native support of OpenTelemetry protocol integration allows you to send collected telemetry data directly from your applications to APM Server.

Reviewer: Maybe one of these instead?

> The APM Server native support of the OpenTelemetry protocol allows you to send collected telemetry data directly from your applications to APM Server.

> The APM Server native OpenTelemetry protocol integration allows you to send collected telemetry data directly from your applications to APM Server.

Author: Good suggestions. I went with the first one.

Trace data collected from your services and the metrics data collected from your applications and infrastructure are sent using the
OpenTelemetry protocol.

image::images/open-telemetry-protocol-arch.png[OpenTelemetry Elastic protocol architecture diagram]

[float]
[[instrument-apps-apm-server]]
===== Instrument applications

To export traces and metrics to APM Server, ensure that you have instrumented your services and applications

> with the OpenTelemetry API, SDK, or both. For example, if you are a Java developer, you need to execute your Java app using the

Reviewer: `execute` again. It's possible this word is correct and I just haven't heard it yet.

Author: `Instrument` is a much better word here.

----
export OTEL_RESOURCE_ATTRIBUTES=service.name=checkoutService,service.version=1.1,deployment.environment=production
export OTEL_EXPORTER_OTLP_ENDPOINT=https://apm_server_url:8200
export OTEL_EXPORTER_OTLP_HEADERS="Authorization=Bearer apm_secret_token"
java -javaagent:/path/to/opentelemetry-javaagent-all.jar \
     -classpath lib/*:classes/ \
     com.mycompany.checkout.CheckoutServiceServer
----

|=== 

| `OTEL_RESOURCE_ATTRIBUTES` | The service name to identify your application.

| `OTEL_EXPORTER_OTLP_ENDPOINT` | APM Server URL. The host and port that APM Server listens for events on.

| `OTEL_EXPORTER_OTLP_HEADERS` | Authorization header that includes the Elastic APM Secret token or API key: `"Authorization=ApiKey api_key"`.

For information on how to format an API key, see our {apm-server-ref-v}/api-key.html[API key] docs.

Please note the required space between `Bearer` and `apm_secret_token`, and `APIKey` and `api_key`.

| `OTEL_EXPORTER_OTLP_CERTIFICATE` | Certificate for TLS credentials of the gRPC client. (optional)

|===

You are now ready to collect <<open-telemetry-elastic-traces-metrics,traces and metrics>>, <<open-telemetry-elastic-verify,verifying metrics>>,

> and <<open-telemetry-elastic-kibana,visualizing metrics>> in {kib}.

Reviewer: Same gerunds as before. Sounds a little off.

export OTEL_EXPORTER_OTLP_ENDPOINT=https://apm_server_url:8200
java -javaagent:/path/to/opentelemetry-javaagent-all.jar \
     -jar target/frontend-1.1.jar
----

|=== 

| `OTEL_RESOURCE_ATTRIBUTES` | The service name to identify your application.

| `OTEL_EXPORTER_OTLP_ENDPOINT` | APM Server URL. The host and port that APM Server listens for events on.

|===

You are now ready to collect <<open-telemetry-elastic-traces-metrics,traces and metrics>>, <<open-telemetry-elastic-verify,verifying metrics>>,
and <<open-telemetry-elastic-kibana,visualizing metrics>> in {kib}.

[[open-telemetry-elastic-protocol]]
==== APM Server native support of OpenTelemetry protocol

experimental::[This feature is experimental and may be changed in a future release. It is only available in a self-managed environment.]

The APM Server native support of OpenTelemetry protocol integration allows you to send collected telemetry data directly from your applications to APM Server.
Trace data collected from your services and the metrics data collected from your applications and infrastructure are sent using the

> OpenTelemetry protocol.

Author: No this is different. This refers to the OpenTelemetry protocol (OTLP).

> OpenTelemetry protocol.

Reviewer: "Elastic's protocol" refers to how Elastic APM Agents send data to a collector, like APM Server. "OTLP" refers to how OpenTelemetry agents send data to a collector, like the OTEL collector.

[float]
[[instrument-apps-apm-server]]
===== Instrument applications

To export traces and metrics to APM Server, ensure that you have instrumented your services and applications
with the OpenTelemetry API, SDK, or both. For example, if you are a Java developer, you need to execute your Java app using the
https://github.com/open-telemetry/opentelemetry-java-instrumentation[OpenTelemetry agent for Java]. 

By defining the following environment variables, you can customize the OTLP endpoint so that the OpenTelemetry agent communicates with
APM Server.

[source,bash]
----
export OTEL_RESOURCE_ATTRIBUTES=service.name=checkoutService,service.version=1.1,deployment.environment=production
export OTEL_EXPORTER_OTLP_ENDPOINT=https://apm_server_url:8200
export OTEL_EXPORTER_OTLP_HEADERS="Authorization=Bearer apm_secret_token"
java -javaagent:/path/to/opentelemetry-javaagent-all.jar \
     -classpath lib/*:classes/ \
     com.mycompany.checkout.CheckoutServiceServer
----

|=== 

| `OTEL_RESOURCE_ATTRIBUTES` | The service name to identify your application.

| `OTEL_EXPORTER_OTLP_ENDPOINT` | APM Server URL. The host and port that APM Server listens for events on.

> | `OTEL_EXPORTER_OTLP_HEADERS` | Authorization header that includes the Elastic APM Secret token or API key. Please note the required space between `Bearer` and `apm-secret-token`.

Reviewer 3: Perhaps provide an example of how to format an API Key too? And/or link to https://www.elastic.co/guide/en/apm/server/current/api-key.html

Reviewer 2: Very good point. I thought about it and dropped the ball. 

Would it be like 
```
export OTEL_EXPORTER_OTLP_HEADERS="Authorization=ApiKey R0gzRWIzUUI3eVpiU054S3pYSy06bXQyQWl4TlZUeEcyUjd4cUZDS0NlUQ=="
```

Reviewer 3: @cyrille-leclerc yes that looks correct.

Re additional headers, I'm onboard with having method-specific headers. I'd still like to wait for feedback like we discussed recently.

Reviewer 2: @EamonnTP can you please document the usage of API akei based authentication with:
``` 
export OTEL_EXPORTER_OTLP_HEADERS="Authorization=ApiKey R0gzRWIzUUI3eVpiU054S3pYSy06bXQyQWl4TlZUeEcyUjd4cUZDS0NlUQ=="
```

@axw are we sure that we can't simplify and also handle the api key authentication with the `OTEL_EXPORTER_OTLP_HEADERS="Authorization=Bearer xxx` format?

Reviewer 3: @cyrille-leclerc are you asking if we can support an additional auth header as well as Authorization? Yes we can. Maybe you could open an issue describing how you would like it to look, and we can discuss further there?

Reviewer 2: I was not clear, my question is: can we use the same `Authorization=Bearer xxx` format for both Secret Token authentication and API Key authentication.
We would support `Authorization=Bearer xxx` with `xxx` being the "secret token" or the "api key"

Reviewer 3: Oh I see. `Authorization: ApiKey <api_key>` was chosen to be consistent with Elasticsearch. We could change that, but that's an even bigger discussion.

Reviewer 2: If it's a bigger discussion, let's park it for the moment :-) We will solve the challenge of clarity primarily with documentation (and hopefully with error messages)

Author: I updated this section. Hopefully it provides more clarity.
